wordpress-mobile/WordPress-Android
# Zendesk changes in preparation of Push Notifications

I've finished the Zendesk Push Notifications in the `issue/zendesk-push-notifications` branch. However, due to some complications we have encountered while testing push notifications, I've had to make a lot of changes that's not directly related to them. To make it easier to review the changes, I've decided to separate it into 2 PRs. The last 2 commits on this PR 175fadf8a3120d970e81953dd272983645129c4e & 21f5da4b121fdc84ddf7da87eb596523d95ff537 removes the push notifications related changes. Once this is merged, I'll open a PR adding those back in.

I'll go over each change in this PR and try to explain them to make it easier to review and understand.

**`SupportHelper` Changes**

`SupportHelper` didn't actually change much. The diff looks kind of weird because of the movement of functions. I've removed `getSupportIdentity` and `getSupportEmailAndNameSuggestion` methods and moved the `showSupportIdentityInputDialog` and `getSupportEmailAndNameSuggestion` static methods under `SupportHelper` class. It also doesn't call the `AppPrefs.setSupportEmail` or `AppPrefs.setSupportName` anymore. This way, the only responsibility of the `SupportHelper` is to provide a way to show an input dialog and let the caller decide what to do with the inputs.

**Injecting `AccountStore` & `SiteStore` to `ZendeskHelper`**

Since we switched away from static methods in #7911 for the `ZendeskHelper` and `SupportHelper`, I can't find a good reason to get them as parameters instead of injecting them directly. As a continuation of this change, I've also opted to pass the minimum required information to the static helpers, so for example, the list of sites is passed instead of the `SiteStore` to `buildZendeskCustomFields`. This was an oversight by me in the original implementation.

Neither of these changes were planned, but happened as part of the Zendesk Identity rework. That's why they weren't in their separate PRs.

**Zendesk Identity changes**

The way I understand it, the Zendesk identity needs to be set before the push notifications can be enabled. This presented a challenge, because we'd get the token at the start of the app where the identity might not be set at all. As far as I know, the only way to check if the identity is set is to check if the `zendeskInstance.identity` is null. However, that only tells us if any identity is set, but it doesn't tell us the email or the name of the identity set. We were previously relying on `AppPrefs.getSupportEmail` and keep updating the identity, but we couldn't tell when the identity was actually changed. Although we can get by without that information, it just made things a lot harder to understand, especially for someone who is unfamiliar with the Zendesk identity works. That's the motivation behind these changes.

Here is how it works now:

* We now have `supportEmail` and `supportName` properties. These will keep track of the email and name set for the Zendesk identity in this session. Having this available makes it possible to set the identity with an anonymous identity (without email and name) and make sure we update it when that's available. So, `setupZendesk` can now set an identity.
* `requireIdentity`, `refreshIdentity` and `clearIdentity` functions introduced. They are documented in the code, so hopefully that's clear enough.
* `setSupportEmail` moved to `ZendeskHelper`, so when the support email is updated from the `HelpActivity`, this function should be called which will also update the Zendesk identity. This will ensure that we always have the correct identity set for Zendesk.
* `showZendeskHelpCenter` doesn't set or refresh the identity anymore. Due to the changes mentioned above, this should no longer be required to do.

------------------------------------------------------

I hate having to make these many changes to a feature that's about to be complete, but the push notification implementation required it and I think it turned out to be better than what it was. So, I guess better late than never.

Please let me know if you have any questions. I am happy to walk anyone through these changes or how Zendesk works.

**To Test:**

I suggest trying to create a Zendesk ticket, verifying it is present in `My Tickets` and visiting the Help Center to check that the faqs are loading correctly. A detailed test is not required for this PR, as it doesn't change the way Zendesk works. We'll also have a much more detailed testing step before this will be merged into `develop` for which I hope to write some detailed testing instructions.

@planarvoid Can I bother you with another review?

/cc @aerych 

_P.S:_ This PR targets a WIP branch, that's why it doesn't have a milestone assigned.
